### PR TITLE
ethtool-util: support more offloading feature settings

### PR DESCRIPTION
**Pull request by yuwata**
~~This is WIP, but comments are welcome. Especially, I'd like to ask opinions about whether we should add all relevant settings to .link or not. If not, which do you want to add?~~

C.f. #20392.

### CODE REVIEW

**Comment by yuwata** [comment]
cc @DaanDeMeyer and @anitazha.

**Comment by poettering** [comment]
> This is WIP, but comments are welcome. Especially, I'd like to ask opinions about whether we should add all relevant settings to .link or not. If not, which do you want to add?

I'd only add those which are obvious that we should have and those people actually ask for.

File: src/shared/ethtool-util.h [hunk 1]
```diff
@@ -20,14 +20,16 @@ typedef enum Duplex {
...
         NET_DEV_FEAT_GSO,
         NET_DEV_FEAT_GRO,
         NET_DEV_FEAT_GRO_HW,
         NET_DEV_FEAT_LRO,
         NET_DEV_FEAT_TSO,
         NET_DEV_FEAT_TSO6,
         _NET_DEV_FEAT_MAX,
+
+        NET_DEV_FEAT_TX = _NET_DEV_FEAT_MAX,
+        _NET_DEV_FEAT_MAX_ALL,
```
Comment (poettering) [inline src/shared/ethtool-util.h]: Can you add a comment explaining why you do this?

And I think `_ALL_MAX` would be a better suffix than `_MAX_ALL`

Comment (poettering) [reply src/shared/ethtool-util.h]: or maybe even _FULL_MAX or _TOTAL_MAX or so?

File: src/shared/ethtool-util.c [hunk 2]
```diff
@@ -434,39 +435,42 @@ int ethtool_set_nic_buffer_size(int *ethtool_fd, const char *ifname, const netde
...
+        uint32_t len;
         int r;
 
         assert(ethtool_fd >= 0);
-        assert(ifr);
+        assert(ifname);
         assert(ret);
 
-        ifr->ifr_data = (void *) &buffer.info;
+        strscpy(ifr.ifr_name, IFNAMSIZ, ifname);
```
Comment (poettering) [inline src/shared/ethtool-util.c]: doesn't matter, but I'd use `sizeof(ifr.ifr_name)` as second arg.

File: src/shared/ethtool-util.c [hunk 3]
```diff
@@ -434,39 +435,42 @@ int ethtool_set_nic_buffer_size(int *ethtool_fd, const char *ifname, const netde
...
 
-        r = ioctl(ethtool_fd, SIOCETHTOOL, ifr);
+        r = ioctl(ethtool_fd, SIOCETHTOOL, &ifr);
```
Comment (poettering) [inline src/shared/ethtool-util.c]: maybe write as:

```c
r = ioctl(ethtool_fd, SIOCETHTOOL, &(struct ifreq) { .ifr_data = &buffer.info });
```

Given that you don't actually care about the ifreq after the ioctl, but just the `buffer` a compound literal sounds good to me.

but really doesn't matter

File: src/shared/ethtool-util.c [hunk 4]
```diff
@@ -476,48 +480,134 @@ static int get_stringset(int ethtool_fd, struct ifreq *ifr, int stringset_id, st
...
+        _cleanup_free_ struct ethtool_gfeatures *gfeatures = NULL;
+        struct ifreq ifr;
+        int r;
 
+        assert(ethtool_fd >= 0);
+        assert(ifname);
+        assert(ret);
+        assert(n_features > 0);
+
+        gfeatures = malloc0(sizeof(struct ethtool_gfeatures) + DIV_ROUND_UP(n_features, 32U) * sizeof(gfeatures->features[0]));
```
Comment (poettering) [inline src/shared/ethtool-util.c]: shouldn't this be:

```c
gfeatures = malloc0(offsetof(struct ethtool_gfeatures, features) + DIV_ROUND_UP(n_features, 32U) * sizeof(gfeatures->features[0]));
```

File: src/shared/ethtool-util.c [hunk 5]
```diff
@@ -476,48 +480,134 @@ static int get_stringset(int ethtool_fd, struct ifreq *ifr, int stringset_id, st
...
+        gfeatures->cmd = ETHTOOL_GFEATURES;
+        gfeatures->size = DIV_ROUND_UP(n_features, 32U);
+
+        ifr = (struct ifreq) {
+                .ifr_data = (void*) gfeatures,
+        };
+        strscpy(ifr.ifr_name, IFNAMSIZ, ifname);
+
+        r = ioctl(ethtool_fd, SIOCETHTOOL, &ifr);
+        if (r < 0)
```
Comment (poettering) [inline src/shared/ethtool-util.c]: you can get rid of `r` actually, and just call ioctl() in the if check itself

File: src/shared/ethtool-util.c [hunk 6]
```diff
@@ -538,26 +628,41 @@ int ethtool_set_features(int *ethtool_fd, const char *ifname, const int features
...
+                return log_debug_errno(r, "ethtool: could not get ethtool feature strings: %m");
 
-        r = get_stringset(*ethtool_fd, &ifr, ETH_SS_FEATURES, &strings);
+        r = get_features(*ethtool_fd, ifname, strings->len, &gfeatures);
         if (r < 0)
-                return log_debug_errno(r, "ethtool: could not get ethtool features for %s", ifname);
+                return log_debug_errno(r, "ethtool: could not get ethtool features for %s: %m", ifname);
+
+        sfeatures = malloc0(sizeof(struct ethtool_sfeatures) + DIV_ROUND_UP(strings->len, 32U) * sizeof(sfeatures->features[0]));
```
Comment (poettering) [inline src/shared/ethtool-util.c]: as above

File: src/shared/ethtool-util.c [hunk 7]
```diff
@@ -538,26 +628,41 @@ int ethtool_set_features(int *ethtool_fd, const char *ifname, const int features
...
-                        r = set_features_bit(strings, netdev_feature_table[i], features[i], sfeatures);
-                        if (r < 0) {
-                                log_debug_errno(r, "ethtool: could not find feature, ignoring: %s", netdev_feature_table[i]);
-                                continue;
-                        }
+        for (size_t i = 0; i < _NET_DEV_FEAT_MAX; i++) {
+                r = set_features_bit(strings, gfeatures, sfeatures, netdev_feature_table[i], features[i]);
+                if (r < 0) {
+                        log_debug_errno(r, "ethtool: could not set feature %s for %s, ignoring: %m", netdev_feature_table[i], ifname);
+                        continue;
```
Comment (poettering) [inline src/shared/ethtool-util.c]: you can drop the continue

File: src/shared/ethtool-util.c [hunk 8]
```diff
@@ -538,26 +628,41 @@ int ethtool_set_features(int *ethtool_fd, const char *ifname, const int features
...
                 }
+        }
 
-        ifr.ifr_data = (void *) sfeatures;
+        for (size_t i = _NET_DEV_FEAT_MAX; i < _NET_DEV_FEAT_MAX_ALL; i++) {
+                r = set_features_multiple_bit(strings, gfeatures, sfeatures, netdev_feature_table[i], features[i]);
+                if (r < 0) {
+                        log_debug_errno(r, "ethtool: could not set feature %s for %s, ignoring: %m", netdev_feature_table[i], ifname);
+                        continue;
```
Comment (poettering) [inline src/shared/ethtool-util.c]: ditto

File: src/shared/ethtool-util.c (hunk 3 above)
Comment (yuwata) [reply src/shared/ethtool-util.c]: No, `ifr.ifr_name` also needs to be set.

**Comment by yuwata** [comment]
Updated. This adds no new settings to .link, and the last two commits are preparations for introducing new offloading settings in the future when people requests some settings. PTAL.

**Comment by yuwata** [comment]
CentOS CI failure is caused by `TEST-15-DROPIN`, unrelated.

**Pull request merged**